README.md
# Dictionary Data Frame

A dictionary implementation of the pandas data frame.

# Contributing

You can submit a pull request or add an issue to track.

# License

This project is license under the GPLv3 license - see the [LICENSE.md](LICENSE.md) file for details


CHANGELOG.md

2018-09-28 v1.0.0:
	- initial release


tests/test_ddf.py
from collections import OrderedDict

import numpy as np
import pandas as pd
import pytest

from ddf import DDF, utils

np.random.seed(123)
N_EXAMPLES = 20

DATA = {
        0: np.random.randn(N_EXAMPLES),
        1: np.arange(N_EXAMPLES),
        2: np.random.choice([1, 2, np.nan], size=N_EXAMPLES),
        'group_id': np.random.choice([1, 2, 3], size=N_EXAMPLES)
        }


@pytest.fixture
def pandas_df():
    pdf = pd.DataFrame(DATA)
    return pdf


@pytest.fixture
def df():
    df = DDF(DATA)
    return df


def test_rename_ignore_missing(df):
    mapping = {'missing_column': 5, 'group_id': 'new_name'}
    df = df.rename(mapping, ignore_missing=True)
    assert 'new_name' in df
    assert 'group_id' not in df


def test_can_pickle(df, tmpdir):
    df['strings'] = np.repeat('a', len(df))
    path = str(tmpdir) + 'df.pkl'
    utils.write_pickle(df, path)
    loaded = utils.read_pickle(path)
    assert loaded.equals(df)
    for col in df:
        assert loaded[col].dtype == df[col].dtype


# TODO: make repr work for this
# df = DDF({'badcol': [u'', u'  ', u'+ ', u'\xa7', u'x ', u'xx', u'd ', u'\xa7\xa7', u'p ', u'? ']})

def test_ddf_repr():
    df = DDF()
    repr(df)


def test_ddf_init():
    data = [{'a': 1}]
    df = DDF(data)
    assert isinstance(df.data['a'], np.ndarray)


def test_fillna_not_inplace(df):
    df[0][:5] = np.nan
    original_df = df.copy(deep=True)
    filler = 999
    new_df = df.fillna(filler)
    assert np.allclose(new_df[0][:5], filler)
    assert np.array_equal(new_df[0][5:], original_df[0][5:])
    assert utils.nan_allclose(df[0], original_df[0])

    filled = df.fillna({0: 101, 2: 999})
    assert np.allclose(filled[0][:5], 101)
    assert np.allclose(filled[2][-4], 999)

    filled = df.fillna({0: np.arange(len(df))})
    assert np.array_equal(filled[0][:5], np.arange(5))


def test_fillna_inplace(df):
    df[0][:5] = np.nan
    original_df = df.copy(deep=True)
    filler = 999
    df.fillna(filler, inplace=True)
    assert np.allclose(df[0][:5], filler)
    assert np.array_equal(df[0][5:], original_df[0][5:])


def test_add_column_which_is_shape_n_1():
    DDF({'col': np.arange(10), 'new_col': np.arange(10).reshape((-1, 1))})


def test_saving_to_csv_by_default_does_not_save_index(tmpdir):
    save_path = str(tmpdir) + '/saved_df.csv'
    df = DDF({'col': np.arange(5)})
    df.to_csv(save_path)
    loaded_df = DDF.from_csv(save_path)
    assert 'Unnamed: 0' not in loaded_df


def test_add_column_to_empty_ddf():
    df = DDF()
    df['a'] = np.arange(5)


def test_set_single_col_with_matrix(df):
    zeros = np.zeros(len(df)).reshape((-1, 1))
    df['a'] = zeros
    assert len(df['a']) == len(df)
    assert np.array_equal(df['a'], zeros.reshape((-1, )))


def test_set_single_col_with_array(df):
    zeros = np.zeros(len(df))
    df['a'] = zeros
    assert len(df['a']) == len(df)
    assert np.allclose(df['a'], zeros)


def test_set_single_col_with_scalar(df):
    df['a'] = 0
    assert len(df['a']) == len(df)
    assert np.allclose(df['a'], 0)


def test_set_multiple_cols_to_same_scalar(df):
    df[['a', 'b']] = 0
    for col in 'a', 'b':
        values = df[col]
        assert np.array_equal(values, np.zeros(len(df)))


def test_set_multiple_existing_cols_to_different_scalars(df):
    cols = df.columns[:2]
    df[cols] = [0, 1]
    assert np.array_equal(df[cols[0]], np.zeros(len(df)))
    assert np.array_equal(df[cols[1]], np.ones(len(df)))


def test_set_multiple_cols_to_matrix(df):
    df[['a', 'b']] = np.zeros((len(df), 2))
    for col in 'a', 'b':
        assert np.array_equal(df[col], np.zeros(len(df)))


def test_adding_columns_to_df_with_one_scalar_row():
    df = DDF({'a': 0.10229005573016618})
    df['b'] = 1


def test_appropriate_setitem(df):
    before = df[1]
    df[1] = np.random.uniform(size=N_EXAMPLES)
    after = df[1]
    assert np.all(before != after)


def test_incorrect_length_setitem(df):
    with pytest.raises(AssertionError):
        df[1] = np.random.randn(10 * N_EXAMPLES)

##########################################
# # __init__
##########################################


def test_empty_init():
    DDF()
    DDF({})


def test_init_with_different_value_shapes():
    data = {'a': 1, 'b': np.arange(2)}
    with pytest.raises(RuntimeWarning):
        DDF(data)


def test_init_with_matrix_values():
    data = {'a': np.arange(10).reshape(5, 2)}
    with pytest.raises(RuntimeWarning):
        DDF(data)

##########################################
# # side effects
##########################################


def test_for_side_effects_outside_df():
    data = {'a': np.arange(5)}
    df = DDF(data)
    data['a'] += 1
    assert np.array_equal(df['a'], np.arange(5))


def test_for_side_effects_in_init_data():
    data = {'a': np.arange(5)}
    data['b'] = data['a']
    df = DDF(data)
    df['a'] += 1
    assert np.array_equal(df['b'], np.arange(5))


def test_for_side_effects_in_added_column():
    df = DDF({})
    df['a'] = np.arange(5)
    df['b'] = df['a']
    df['a'] += 1
    assert np.array_equal(df['b'], np.arange(5))

##########################################
# # Appending
##########################################


def test_appending_dataframes(df):
    new_df = df.append(df)
    assert len(new_df) == 2 * len(df)
    assert new_df.columns == df.columns

    new_col = df.colslice(0)
    new_col.rename({0: 99}, inplace=True)
    new_df = df.append(new_col)
    assert len(new_df) == len(df)
    assert 99 in new_df


def test_appending_with_timedeltas():
    df = DDF({'delta': np.array([100, 'NaT'], dtype='timedelta64[ns]')})
    other_df = DDF()
    df.append(other_df, axis=0)


def test_appending_dfs_with_one_scalar_row():
    df = DDF({'a': 0.10229005573016618})
    other_df = DDF({'b': 0.10229005573016618})
    df.append(other_df, axis=0)


def test_appending_new_df_with_new_cols():
    d1 = DDF({'col1': np.arange(4), 'col2': np.arange(4)})
    d2 = DDF({'col1': np.arange(2), 'col3': np.arange(2)})
    d3 = d1.append(d2, axis=0)
    assert len(d3) == 6


def test_concatenate_df_with_non_overlapping_cols():
    d1 = DDF({'col1': np.arange(4), 'col2': np.arange(4)})
    d2 = DDF({'col1': np.arange(2)})
    d3 = d1.append(d2, axis=0)
    expected_col1 = np.array([0, 1, 2, 3, 0, 1])
    expected_col2 = np.array([0, 1, 2, 3, np.nan, np.nan])
    np.testing.assert_array_equal(d3['col1'], expected_col1)
    np.testing.assert_array_equal(d3['col2'], expected_col2)


def test_appending_not_inplace(df):
    original_df = df.copy(deep=True)
    temp_df = df.rename(lambda col: str(col) + '_suffix')
    df.append(temp_df, axis=1)
    assert df.equals(original_df)
    df.append(temp_df, axis=0)
    assert df.equals(original_df)


def test_cannot_append_wrong_size(df):
    with pytest.raises(ValueError):
        df.append(df[:2], axis=1)


def test_appending_empty_df(df):
    df.append(DDF())

##########################################
# # Slicing
##########################################


def test_boolean_array_slicing(df):
    index = np.random.choice([True, False], len(df))
    sliced = df.rowslice(index)
    assert sliced.equals(df[index])


def test_direct_slicing(df):
    assert df[:5].equals(df.head())
    assert df[-5:].equals(df.tail())
    reverse_ix = np.arange(len(df))[::-1]
    assert df[::-1].equals(df.rowslice(reverse_ix))


def test_cant_slice_with_wrong_size_index(df):
    index = np.repeat(True, 10)
    with pytest.raises(AssertionError):
        df[index]

##########################################
# # Other
##########################################


def test_merge_preservers_strings():
    d1 = DDF({'a': np.arange(4), 'b': np.repeat('b', 4), 'd': np.repeat('r', 4)})
    d2 = DDF({'a': np.arange(4), 'c': np.repeat('c', 4), 'd': np.repeat('2r', 4)})
    merged = d1.merge(d2, on='a')
    assert merged['b'].dtype.type is np.str_
    assert merged['c'].dtype.type is np.str_


def test_can_merge_two_ddfs():
    d1 = DDF({'col1': np.arange(4), 'col2': np.arange(4)})
    d2 = DDF({'col1': np.arange(2), 'col3': np.arange(2)})
    d3 = d2.merge(d1, on='col1', how='outer')
    assert np.allclose(d3['col2'], d1['col2'])
    assert utils.nan_allclose(d3['col3'], np.array([0, 1, np.nan, np.nan]))


def test_set_col(df):
    df['col'] = 0
    assert np.all(df['col'] == np.repeat(0, len(df)))


def test_columns(df, pandas_df):
    df_cols, ddf_cols = pandas_df.columns, df.columns
    assert all(df_col == ddf_col for df_col, ddf_col in zip(df_cols, ddf_cols))


def test_values(df, pandas_df):
    np.testing.assert_allclose(pandas_df.values, df.values)


def test_setitem_scalar_to_single_column(df):
    df['test'] = 123.0
    assert np.all(df['test'] == np.repeat(123.0, len(df)))


def test_setitem_scalar_to_multiple_columns(df):
    df[['test0', 'test1']] = 321.0
    shape = (len(df), 2)
    target_array = np.repeat(321.0, np.prod(shape)).reshape(shape)
    assert np.all(df[['test0', 'test1']] == target_array)


def test_setitem_list_to_single_column(df):
    df['test'] = range(len(df))
    assert np.all(df['test'] == np.arange(len(df)))


def test_setitem_list_to_multiple_column(df):
    list_to_set = [[0, 1]] * len(df)
    df[['test0', 'test1']] = list_to_set
    assert np.all(df[['test0', 'test1']] == np.array(list_to_set))


def test_sort(df):
    df['noise'] = np.random.randn(len(df))
    sorted_noise = np.sort(df['noise'])
    assert not np.allclose(df['noise'], sorted_noise)
    sorted_ddf = df.rowsort('noise')
    assert np.allclose(sorted_ddf['noise'], sorted_noise)
    sorted_ddf = df.rowsort('noise', ascending=False)
    assert np.allclose(sorted_ddf['noise'], sorted_noise[::-1])


def test_equals_returns_false_when_columns_are_not_the_same():
    df = DDF({'col1': np.array([1, 2])})
    df2 = DDF({'col2': np.array([1, 2])})
    assert not df.equals(df2)


def test_equals_invariant_to_column_order():
    df = OrderedDict([
            ('col1', np.array([1, 2])),
            ('col2', np.array([2, 1])),
            ])
    df = DDF(df)
    df2 = OrderedDict([
            ('col2', np.array([2, 1])),
            ('col1', np.array([1, 2])),
            ])
    df2 = DDF(df2)
    assert df.equals(df2)


def test_shallow_copy(df):
    copied = df.copy(deep=False)
    copied['new_col'] = np.arange(len(copied))
    assert 'new_col' not in df


def test_drop_rows(df):
    df['rownr'] = np.arange(len(df))
    rows_to_drop = [1, 5, 10, 15]
    new_df = df.drop_rows(rows_to_drop)
    expected_rownrs = np.array([0, 2, 3, 4, 6, 7, 8, 9, 11, 12, 13, 14, 16, 17, 18, 19])
    assert np.array_equal(new_df['rownr'], expected_rownrs)


def test_colslice_with_filter(df):
    def filter_func(col):
        return str(col).endswith('id')
    new_df = df.colslice(filter_func)
    assert new_df.columns == ['group_id']


def test_renaming_inplace(df):
    df.rename({0: 'renamed_0'}, inplace=True)
    assert 'renamed_0' in df
    assert 0 not in df
    df.rename(lambda col: str(col), inplace=True)
    assert '1' in df
    assert 1 not in df


def test_renaming_not_inplace(df):
    df = df.copy(deep=True)
    new_df = df.rename({0: 'renamed_0'})
    assert 'renamed_0' in new_df
    assert 0 not in new_df
    assert 0 in df
    new_df = df.rename(lambda col: str(col))
    assert '1' in new_df
    assert 1 not in new_df
    assert 1 in df


def test_to_dict(df):
    dct = df.to_dict()
    assert dict(df._data) == dct


def test_colapply(df):
    def replace_with_zero(values):
        return np.zeros(len(values))
    df.colapply(0, replace_with_zero)
    assert np.allclose(df[0], 0)

    original = df.copy(deep=True)
    new_df = df.colapply(0, replace_with_zero, inplace=False)
    assert np.allclose(new_df[0], 0)
    assert np.allclose(original[0], df[0])


def test_drop_duplicates_does_not_change_dtypes(df):
    df['a'] = np.repeat('a', len(df))
    df = df.append(df)
    df = df.drop_duplicates()
    assert df['a'].dtype.type is np.str_


def test_merge_unicode_col():
    bad_df = DDF(DATA)
    bad_str = u'hurac\xe1n'
    bad_df['name'] = np.array([bad_str]*N_EXAMPLES).astype(unicode)
    good_df = DDF(DATA)
    good_str = u'namey_mc_name'
    good_df['name'] = np.array([good_str]*N_EXAMPLES).astype(str)
    bad_df.merge(good_df, on=['group_id', 'name'], how='left')


def test_to_json(df):
    with pytest.raises(AssertionError):
        df.to_json()
    df = df.rename(lambda col: str(col))
    df = df.sort('0')
    json = df.to_json()
    newdf = DDF.from_json(json)
    newdf = newdf.sort('0')
    assert newdf.equals(df)


def test_json_conversion_datetimes():
    dates = np.array([
        '2013-03-30T00:00:00.000000000', '2014-02-10T00:00:00.000000000',
        '2014-05-12T00:00:00.000000000', '2014-03-17T00:00:00.000000000',
        '2013-04-12T00:00:00.000000000', '2014-05-19T00:00:00.000000000',
        '2014-04-09T00:00:00.000000000', '2014-05-11T00:00:00.000000000',
        '2014-02-25T00:00:00.000000000', '2014-04-24T00:00:00.000000000',
        '2014-05-30T00:00:00.000000000', '2014-02-09T00:00:00.000000000',
        '2014-05-05T00:00:00.000000000'], dtype='<M8[ns]')
    df = DDF({'time': dates})
    js = df.to_json()
    newdf = DDF.from_json(js)
    newdf = newdf.colslice(df)
    assert newdf.equals(df)


def test_json_conversion_strings_with_leading_zeros():
    runner_ids = np.array(['000000000' + str(x) for x in range(10)])
    df = DDF({'runner_id': runner_ids})
    js = df.to_json()
    newdf = DDF.from_json(js)
    newdf = newdf.colslice(df)
    assert newdf.equals(df)


def test_json_conversion_preserves_string_dtypes():
    original_df = DDF({
        'runner_id': np.array(['test', 'testing'], dtype='S7'),
        'short_strings': np.array(['a', 'b'], dtype='S1'),
        'more_strings': np.array(['abc', 'def'], dtype='S')
    })

    js = original_df.to_json()
    new_df = DDF.from_json(js)
    new_df = new_df.colslice(original_df)
    assert new_df.equals(original_df)

    for col in ['runner_id', 'short_strings', 'more_strings']:
        assert new_df[col].dtype == original_df[col].dtype


ddf/core.py
"""
DDF
Copyright (C) 2018 Seamless Global Ltd.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.

"""

from collections import OrderedDict, defaultdict
from functools import wraps
import copy
import json
import warnings

from future.utils import raise_with_traceback
import numpy as np
import pandas as pd

from ddf import utils

default_config = {
            'repr': {
                'style': 'horizontal',
                'max_length': 10
                }
            }


class DDF(object):
    def __init__(self, data=None):
        """
        Tabular data sctructure, essentially a dictionary with added functionality.

        Parameters
        ----------
        data (dict, list, other): if data type is not dict of list, it must be compatible
            with the construction of a OrderedDict

        Examples
        --------
        >>> df = DDF()
        >>> df
            Empty DDF
            -----
            Shape: (0, 0)
        >>> df = DDF({'a': [1,2,3], 'b': [2,3,4]})
        >>> df
                a   b
            0   1   2
            1   2   3
            2   3   4

            -----
            Shape: (3, 2)
        >>> df = DDF([{'a':1}, {'a':2}])
        >>> df
                a
            0   1
            1   2

            -----
            Shape: (2, 1)
        """
        data = copy.copy(data)
        if data is None:
            data = OrderedDict()
        elif isinstance(data, dict):
            data = self._parse_dict(data)
        elif isinstance(data, list):
            data = self._convert_list_to_dict(data)
            data = self._parse_dict(data)
        else:
            try:
                data = OrderedDict(data)
            except Exception:
                raise TypeError('cant instantiate from {}'.format(type(data)))
        self._data = data

        self._config = default_config

    def colapply(self, cols, func, inplace=True):
        def _colapply(cols, funcs, data):
            for col in cols:
                data[col] = func(data[col])

        cols = utils.ensure_is_list(cols)
        if inplace:
            _colapply(cols, func, self._data)
        else:
            new_data = copy.deepcopy({k: self[k] for k in cols})
            new_data.update(copy.copy({k: self[k] for k in self if k not in cols}))
            _colapply(cols, func, new_data)
            result = type(self)(new_data)
            return result

    def rename(self, renamer, inplace=False, ignore_missing=False):
        """
        Rename columns of df.

        Arguments
        ---------
            renamer (dict, func): transformations to apply to columns

        Returns
        -------
            renamed_df (df): new object

        Examples
        --------
        >>> df = DDF({'a': range(10)})
        >>> df.rename(lambda col: col + '_suffix', inplace=True)
        >>> df.rename({'a_suffix': 'b'}, inplace=True)
        """
        if not isinstance(renamer, dict):
            renamer = {c: renamer(c) for c in self}

        def _rename_columns(dct):
            for old, new in renamer.iteritems():
                try:
                    dct[new] = dct.pop(old)
                except KeyError as e:
                    if not ignore_missing:
                        raise_with_traceback(e)

        if inplace:
            _rename_columns(self._data)
        else:
            new_data = self._data.copy()
            _rename_columns(new_data)
            new_df = type(self)(new_data)
            return new_df

    def fillna(self, value, inplace=False):
        """
        Replacing null values in df.

        Arguments
        ---------
            value (scalar, dictionary): if scalar, the same value will be used to fill
                all null values in the df. If a dictionary, the holes will be filled with
                the values specified for each column. If the filling value are an
                `np.ndarray`, the array will be filled with the fill value at the
                corresponding index.

        Returns
        -------
            filled (df)

        Example
        -------
        >>> df = DDF({'a': np.array([1, 2, 3, 4, None, np.nan, 7, 8, 9, 10])})
        >>> df.fillna(999)
        >>> df.fillna({0:np.arange(len(df))})
        >>> df['b'] = np.array([1, 2, 3, 4, 5, 6, np.nan, np.nan, np.nan, np.nan])
        >>> df.fillna({'a':101, 'b':999})
        """
        warnings.warn('Fillna is not longer in place by default!')

        def _fillna(dct, value):
            if isinstance(value, dict):
                for colname, values in dct.items():
                    if colname in value:
                        fillvalue = value[colname]
                        fillix = pd.isnull(values)
                        if isinstance(fillvalue, np.ndarray):
                            values[fillix] = fillvalue[fillix]
                        else:
                            values[fillix] = fillvalue
            else:
                for v in dct.itervalues():
                    v[pd.isnull(v)] = value

        if inplace:
            _fillna(self._data, value)
        else:
            cols_to_copy = [k for k, v in self._data.iteritems() if any(pd.isnull(v))]
            new_data = copy.deepcopy({k: self[k] for k in cols_to_copy})
            new_data.update(copy.copy({k: self[k] for k in self if k not in cols_to_copy}))
            _fillna(new_data, value)
            filled = type(self)(new_data)
            return filled

    def iterrows(self):
        n_rows = len(self)
        for row in xrange(n_rows):
            yield self.rowslice(row)

    def pop(self, column):
        return self._data.pop(column)

    def drop_rows(self, row_numbers):
        ix = utils.ix_to_bool(row_numbers, len(self))
        return self[~ix]

    def head(self, N=5):
        return self.rowslice(np.arange(N))

    def tail(self, N=5):
        tail_ix = np.arange(len(self) - N, len(self))
        return self.rowslice(tail_ix)

    @classmethod
    def from_pandas(cls, df):
        data = OrderedDict((column, df[column].values) for column in df)
        return cls(data)

    @classmethod
    def from_hdf(cls, path, key='df'):
        df = pd.read_hdf(path, key)
        return cls.from_pandas(df)

    @classmethod
    def from_csv(cls, path, **kwargs):
        df = pd.read_csv(path, **kwargs)
        return cls.from_pandas(df)

    def to_pandas(self):
        return pd.DataFrame(self._data)

    def to_hdf(self, path, key='df'):
        df = self.to_pandas()
        df.to_hdf(path, key)

    def to_csv(self, path, **kwargs):
        df = self.to_pandas()
        if 'index' not in kwargs:
            kwargs['index'] = False
        df.to_csv(path, **kwargs)

    def to_json(self):
        assert all(map(lambda c: isinstance(c, (str, unicode)), self.columns))
        df = self.to_pandas()
        raw_json = df.to_json(date_format='iso')
        dtypes = {c: str(self[c].dtype) for c in df}
        js = {'values': raw_json, 'dtypes': dtypes}
        js = json.dumps(js)
        return js

    @classmethod
    def from_json(cls, jsons):
        js = json.loads(jsons)
        df = pd.read_json(js['values'], dtype=js['dtypes'])
        df = df.sort_index()
        df = cls.from_pandas(df)
        for col, col_type in js['dtypes'].items():
            df[col] = df[col].astype(col_type)
        return df

    def rowslice(self, index):
        row_slices = OrderedDict((k, v[index]) for k, v in self._data.iteritems())
        return type(self)(row_slices)

    def colslice(self, columns_or_filter):
        """
        Creating a new df with selected columns

        Arguments
        ---------
            slicer (list, filter_function): if `list`, the new ddf will have the columns
                in the list. If filter_function, a list will be created according to
                the columns selected by the filter.

        Examples
        --------
        >>> df = DDF({'a': [1,2,3], 'b': [1,2,3]})
        >>> df.colslice(lambda col: col == 'a')
                a
            0   1
            1   2
            2   3

            -----
            Shape: (3, 1)
        >>> df.colslice(['b'])
                b
            0   1
            1   2
            2   3

            -----
            Shape: (3, 1)
        """
        def is_iterator(obj):
            return hasattr(obj, 'next')
        if callable(columns_or_filter) and (not is_iterator(columns_or_filter)):
            selected = filter(columns_or_filter, self.columns)
        elif np.isscalar(columns_or_filter):
            selected = utils.ensure_is_list(columns_or_filter)
        else:
            selected = columns_or_filter
        col_slice = OrderedDict((k, self.data[k]) for k in selected)
        return self.__class__(col_slice)

    def sort(self, keys, axis=0, ascending=True):
        if axis == 0:
            return self.rowsort(keys, ascending)
        else:
            raise NotImplementedError

    def rowsort(self, keys, ascending=True):
        # TODO: inplace?
        if np.isscalar(keys):
            keys = [keys]
        sort_ix = np.lexsort([self.data[k] for k in keys[::-1]])
        if not ascending:
            sort_ix = sort_ix[::-1]
        return self.rowslice(sort_ix)

    @property
    def columns(self):
        return self._data.keys()

    @property
    def values(self):
        return np.array(self._data.values()).T

    def copy(self, deep=True):
        if deep:
            copied = copy.deepcopy(self)
        else:
            copied = copy.copy(self)
            data = copy.copy(self._data)
            self._data = data
        return copied

    def append(self, other_df, axis=None):
        """
        Appending other DDFs to self. It is not always necessary to provide the axis
        for appending - often this method will understand which axis is supposed to be
        used.

        Arguments
        ---------
            other_df (ddf): df to append
            axis (int): axis over which to append
        """
        if len(self) == 0:
            appended = other_df
            return appended
        if len(other_df) == 0:
            return self

        if axis is None:
            has_strictly_new_cols = len(set(other_df).difference(self)) == len(set(other_df))
            if sorted(other_df) == sorted(self):
                appended = self._append_rows(other_df)
            elif len(other_df) == len(self) and has_strictly_new_cols:
                appended = self._append_columns(other_df)
            else:
                raise RuntimeError('appending is ambiguous; please provide axis.')
        else:
            if axis == 0:
                appended = self._append_rows(other_df)
            elif axis == 1:
                appended = self._append_columns(other_df)
            else:
                raise NotImplementedError('DDF only has two axes: 0 and 1.')
        return appended

    @property
    def shape(self):
        return (len(self), len(self._data))

    @wraps(pd.merge)
    def merge(self, other_df, **kwargs):
        def is_string(dtype):
            return (dtype.type is np.str_) or (dtype.type is np.unicode_)

        def get_stringtypes(df, stringtypes=None):
            if stringtypes is None:
                stringtypes = defaultdict(set)
            for col in df:
                dtype = df[col].dtype
                if is_string(dtype):
                    stringtypes[col].add(dtype)
            return stringtypes

        strings = get_stringtypes(self)
        strings = get_stringtypes(other_df, strings)

        self_df = self.to_pandas()
        if hasattr(other_df, 'to_pandas'):
            other_df = other_df.to_pandas()

        merged = self_df.merge(other_df, **kwargs)
        merged = DDF.from_pandas(merged)
        for col, dtypes in strings.items():
            if col not in merged:
                continue
            newdtype = dtypes.pop() if len(dtypes) == 1 else np.result_type(*dtypes)
            if newdtype.type is np.str_:
                merged[col] = merged[col].astype(str)
            elif newdtype.type is np.unicode_:
                merged[col] = merged[col].astype(unicode)
        return merged

    def drop_duplicates(self, *args, **kwargs):
        string_cols = [f for f in self if self[f].dtype.type is np.str_]
        self_df = self.to_pandas()
        new_df = self_df.drop_duplicates(*args, **kwargs)
        df = self.from_pandas(new_df)
        df.colapply(string_cols, lambda v: v.astype(str))
        return df

    def to_dict(self):
        return dict(self._data)

    def _parse_dict(self, data):
        if len(data) == 0:
            data = OrderedDict()
        else:
            if isinstance(data, dict):
                for k, v in data.iteritems():
                    if not isinstance(v, np.ndarray):
                        if isinstance(v, list) and np.any([isinstance(_, np.datetime64) for _ in v]):
                            v = pd.to_datetime(v).values
                        else:
                            v = np.atleast_1d(np.array(v))
                    if v.ndim == 2 and v.shape[1] == 1:
                        v = v.flatten()
                    data[k] = copy.copy(v)
            data = OrderedDict(data)
            self._check_data_format(data)
        return data

    def _convert_list_to_dict(self, data):
        def is_list_of_dicts(data):
            return isinstance(data, list) and all([isinstance(x, dict) for x in data])

        if is_list_of_dicts:
            data = utils.concatenate_dicts(data, np.nan)
        return data

    @property
    def data(self):
        return self._data

    @data.setter
    def data(self, new_data):
        # TODO: fix order of operations
        self._data = new_data
        self._check_internal_data_dict_format()

    def __getitem__(self, indexer):
        def _is_boolean_array(v):
            bools = (bool, np.bool_)
            return isinstance(v, np.ndarray) and all([isinstance(x, bools) for x in v])

        if isinstance(indexer, (list, tuple)):
            columns = indexer
            return np.array([self._data[column] for column in columns]).T
        elif isinstance(indexer, slice):
            index = indexer
            return self.rowslice(index)
        elif _is_boolean_array(indexer):
            assert len(indexer) == len(self)
            return self.rowslice(indexer)
        else:
            key = indexer
            return self._data[key]

    def __setitem__(self, columns, values):
        columns, values = self._ensure_setitem_args_are_valid(columns, values)
        for column, value in zip(columns, values.T):
            self._data[column] = copy.copy(value)
        self._check_internal_data_dict_format()

    def _ensure_setitem_args_are_valid(self, columns, values):
        columns = utils.ensure_is_list(columns)
        values = self._ensure_values_is_matrix(values, columns)
        n_rows, n_cols = values.shape
        if len(self) > 0:
            assert len(self) == n_rows, 'Array does not have the correct number of rows.'
        assert len(columns) == n_cols, 'Array does have the correct number of columns.'
        return columns, values

    def _ensure_values_is_matrix(self, values, columns):
        if isinstance(values, list):
            values = np.array(values)
        elif np.isscalar(values):
            target_shape = (len(self), len(columns))
            values = np.repeat(values, np.prod(target_shape)).reshape(target_shape)
        if values.ndim == 1:
            if len(values) == len(columns):
                values = np.broadcast_to(values, (len(self), len(columns)))
            else:
                values = values.reshape(-1, 1)
        return values

    def __len__(self):
        if len(self._data) == 0:
            return 0
        else:
            first_value = self._data.itervalues().next()
            return 1 if np.isscalar(first_value) else len(first_value)

    def __repr__(self):
        formatter = HorizontalFormatter(max_length=10)
        return formatter.get_repr(self)

    def __iter__(self):
        return self._data.iterkeys()

    def _append_rows(self, other_df):
        all_columns = set(other_df.columns + self.columns)
        len_df, len_self = len(other_df), len(self)
        data = self._data.copy()
        for column in all_columns:
            u = data.get(column, np.repeat(np.nan, len_self))
            u = np.atleast_1d(u)
            try:
                v = other_df[column]
            except KeyError:
                if utils.is_nptimedelta(u):
                    v = np.repeat(np.nan, len_df).astype(u.dtype)
                else:
                    v = np.repeat(np.nan, len_df)
            v = np.atleast_1d(v)
            u, v = self._reconcile_dtypes(u, v)
            uv = np.concatenate([u, v])
            data[column] = uv
        new_df = type(self)(data)
        return new_df

    def _reconcile_dtypes(self, u, v):
        # TODO: how to concatenate arbitrary vectors? May set priority:
        # you always try to cast the other DF's array into your own
        # type?
        if utils.is_datetime(u) or utils.is_datetime(v):
            dtype = u.dtype if utils.is_datetime(u) else v.dtype
            return u.astype(dtype), v.astype(dtype)
        else:
            return u, v

    def _append_columns(self, other_df):
        if len(other_df) != len(self):
            raise ValueError('dfs need to be of the same length')
        data = self._data.copy()
        for column, array in other_df._data.iteritems():
            data[column] = array
        new_df = type(self)(data)
        return new_df

    def equals(self, other_df):
        def get_comparison_func(value):
            if value.dtype in [np.float, np.float32]:
                return np.allclose
            else:
                return np.array_equal

        if len(self) != len(other_df):
            return False
        if sorted(self.columns) != sorted(other_df.columns):
            return False
        for col, value in self._data.iteritems():
            other_value = other_df[col]
            if 'datetime' in str(value.dtype):
                all_equal = (other_value == value).all()
            else:
                skip_ix = pd.isnull(value) & pd.isnull(other_value)
                compare_ix = ~skip_ix
                comparison_func = get_comparison_func(value)
                all_equal = comparison_func(value[compare_ix], other_value[compare_ix])
            if not all_equal:
                return False
        return True

    def _check_internal_data_dict_format(self):
        self._check_data_format(self._data)

    def _check_data_format(self, dct):
        if not all(type(v) == np.ndarray for v in dct.itervalues()):
            raise RuntimeWarning('Not all values are NumPy array. ')

        unique_ndims = np.unique([v.ndim for v in dct.itervalues()])
        if len(unique_ndims) > 1:
            raise RuntimeWarning('Arrays do not have the same dimensions. ')
        if unique_ndims[0] != 1:
            raise RuntimeWarning('Arrays are not one dimensional. ')

        lengths = np.unique([len(v) for v in dct.itervalues()])
        if len(lengths) > 1:
            raise RuntimeWarning('Arrays do not have the same lengths. ')

    def _ipython_key_completions_(self):
        return self.columns


class HorizontalFormatter(object):
    def __init__(self, max_length=10):
        self.max_length = max_length

    def get_repr(self, df):
        items = self._get_repr_item_generator(df)
        table = '\n'.join([self._convert_col_to_str(c, v) for c, v in items])
        shape = 'DDF Shape: {}'.format(df.shape)
        return '{}\n-----\n{}'.format(table, shape)

    def _get_repr_item_generator(self, df):
        def is_first_or_last(j):
            return (j < self.max_length) or (j >= (num_columns - self.max_length))

        num_columns = len(df.columns)
        for j, (column, value) in enumerate(df.data.iteritems()):
            if is_first_or_last(j):
                yield column, value
            elif (j == self.max_length):
                yield '\n.....', '.....\n'

    def _convert_col_to_str(self, column, value):
        if isinstance(value, np.ndarray) and value.dtype == '<M8[ns]':
            value = value.astype('datetime64[D]')
        column = str(column)
        if len(column) > 80:
            column = column[:77] + '...'
        value = str(value).replace('\n', '\n\t')
        return '{}:\n\t{}'.format(column, value)
